### DISCUSSION TITLE
Is there a way to preserve ?_rsc query params on redirection with ISR  on app router? (because Missing ?_rsc params cause CDN issue)

### DISCUSSION THREAD

**Discussion by zecka** [vercel/next.js#65783] [App Router]
I use redirection in some pages with `import { permanentRedirect } from 'next/navigation';`

Once the website is build redirection work well but cause issue with CDN caching because of missing ?_rsc query params. (In my case CDN is Azure front door)

For example i have following redirection

website.com/news/my-article-old-url --> website.com/news/my-article-new-url

On prefetch or on router page change nextjs execute following request `website.com/news/my-article-old-url?_rsc=hash` with a header request Rsc=1.

This request is redirect from `website.com/news/my-article-old-url?_rsc=hash` to `website.com/news/my-article-new-url` and then CDN cache the rsc result as page result, so if we reload the page website.com/news/my-article-new-url we got json like response instead of html.

I already try to intercept request with header Rsc=1 and without queryParams ?_rsc on middleware to add response header Cache-Control: no-cache. But _rsc query params and rsc header is always null on middleware...

**Comment by magnnus** [comment]
I had encounted the same issue. But i only use the step1.

Step 1. remove all internal redirect ( triggered by user click)

Below you can have a try.

Step2. delete loading.js (DO NOT use streaming render  - always 200 http status code)
Step3. have a custom server ( nginx)
Step4. rewrite response cache-control(no-cache) in nginx if the proxied server(next.js) has a 3xx status code AND request header has a  `rsc` field

**Reply by magnnus** [reply]
if the request has a `rsc` request header field AND the request url has no args `?rsc=...` , then set `Cache-Control: private, max-age=0, no-cache, no-store, must-revalidate` in Nginx.

### e.g: Nginx njs snippets

```
if (r.headersIn.rsc && !r.variables.arg_rsc) {
    r.headersOut['Cache-Control'] = `private, max-age=0, no-cache, no-store, must-revalidate`;
 }
```

**Although it can solve the problem, we still hope that the official can provide a more concise solution**

**Reply by magnnus** [reply]
if CDN supports cache key (Vary),  It should be no problem

**Comment by magnusriga** [comment]
@zecka Did you figure out how to spot the request in middleware? In our case, we have redirect in middleware, but see a lot of errors on rsc requests failing after redirect, presumably because redirect drops the search params.

Thanks a lot for any advice!
